mono/profiler/log.c

```diff
@@ -637,31 +640,48 @@ buffer_lock (void)
 	 * is about to stop.
 	 */
 	if (mono_atomic_load_i32 (&log_profiler.buffer_lock_state) != get_thread ()->small_id << 16) {
-		MONO_ENTER_GC_SAFE;
+		/* We can get some sgen events (for example gc_handle_deleted)
+		 * from threads that are unattached to the runtime (but that
+		 * are attached to the profiler).  In that case, avoid mono
+		 * thread state transition to GC Safe around the loop, since
+		 * the thread won't be participating in Mono's suspension
+		 * mechianism anyway.
+		 */
+		MonoThreadInfo *info = mono_thread_info_current_unchecked ();
+		if (info) {
+			MONO_ENTER_GC_SAFE_WITH_INFO (info);
 
-		gint32 old, new_;
+			buffer_lock_helper ();
 
-		do {
-		restart:
-			// Hold off if a thread wants to take the exclusive lock.
-			while (mono_atomic_load_i32 (&log_profiler.buffer_lock_exclusive_intent))
-				mono_thread_info_yield ();
+			MONO_EXIT_GC_SAFE_WITH_INFO;
+		} else
+			buffer_lock_helper ();
+	}
 
-			old = mono_atomic_load_i32 (&log_profiler.buffer_lock_state);
+	mono_memory_barrier ();
+}
 
-			// Is a thread holding the exclusive lock?
-			if (old >> 16) {
-				mono_thread_info_yield ();
-				goto restart;
-			}
+static void
+buffer_lock_helper (void)
+{
+	gint32 old, new_;
 
-			new_ = old + 1;
-		} while (mono_atomic_cas_i32 (&log_profiler.buffer_lock_state, new_, old) != old);
+	do {
+	restart:
+		// Hold off if a thread wants to take the exclusive lock.
+		while (mono_atomic_load_i32 (&log_profiler.buffer_lock_exclusive_intent))
+			mono_thread_info_yield ();
```

Comment:
OK do mono_threads_platform_yield without being in safe mode? It looks like we do it in other places and probably should be OK, but will involve a system call that now could potentially be done from a thread not in safe mode.